### PR TITLE
[Spec Decode] Add Sliding Window Attention support to DFlash drafter

### PR DESCRIPTION
**Pull request by jianc99**
## Purpose

Adds Sliding Window Attention (SWA) support to the DFlash speculative decoding drafter so DFlash draft models with mixed `sliding_attention` / `full_attention` layers can draft correctly.

Without this, SWA layers in the drafter lose their windowed-attention configuration and run as full attention, which hurts acceptance length on long-context inputs. This version is rebased onto current `main` and keeps the PR focused on generic DFlash/SWA infrastructure.

## Changes

- Preserve DFlash SWA config fields when extracting the speculators HF-format config.
- Build DFlash draft attention layers with the correct `sliding_window` setting from `layer_types`.
- Keep DFlash SWA visible to attention metadata while allocating full draft KV for those layers, so target-prewritten context K/V is not evicted by masked draft-block tokens.
- Carry per-KV-group DFlash metadata so full-attention and SWA draft groups can coexist cleanly.
- Use per-KV-group DFlash block tables and slot mappings while keeping the target/draft raw KV tensor shared, preserving effective KV capacity.
- Add a small raw-KV layout bridge so shared DFlash K/V remains correct when target and draft attention backends expose different cache layouts.
- Avoid the EAGLE cache-drop path for DFlash, where the draft context K/V is pre-written by the target model.
- Normalize DFlash auxiliary layer IDs consistently: checkpoint-native `dflash_config.target_layer_ids` follow HF DFlash semantics and are shifted by `+1` when converted to vLLM aux-hidden-state indices; already shifted `eagle_aux_hidden_state_layer_ids` are used as-is.
- Include the normalized auxiliary layer IDs in the compile hash so changing DFlash target layer selection invalidates the compiled graph correctly.

## Test Plan

Focused unit tests:

```bash
PATH=/home/zlab/miniconda3/envs/vllm-dflash/bin:$PATH python -m pytest \
  tests/v1/core/test_kv_sharing.py \
  tests/v1/worker/test_gpu_model_runner.py::test_kv_major_cache_can_share_block_major_raw_tensor \
  tests/v1/spec_decode/test_dflash_swa.py -q
```

Syntax/whitespace hygiene:

```bash
PATH=/home/zlab/miniconda3/envs/vllm-dflash/bin:$PATH python -m py_compile \
  tests/v1/spec_decode/test_dflash_swa.py \
  tests/v1/core/test_kv_sharing.py \
  tests/v1/worker/test_gpu_model_runner.py \
  vllm/config/speculative.py \
  vllm/model_executor/models/qwen3_dflash.py \
  vllm/transformers_utils/configs/speculators/algos.py \
  vllm/v1/core/kv_cache_utils.py \
  vllm/v1/core/sched/scheduler.py \
  vllm/v1/spec_decode/dflash.py \
  vllm/v1/spec_decode/llm_base_proposer.py \
  vllm/v1/worker/gpu_model_runner.py

git diff --check origin/main...HEAD
git diff --check
```

Real model verification:

```bash
vllm serve Qwen/Qwen3.5-122B-A10B \
  --tensor-parallel-size 4 \
  --speculative-config '{"model":"z-lab/Qwen3.5-122B-A10B-DFlash","method":"dflash","num_speculative_tokens":15}' \
  --attention-backend flash_attn \
  --max-num-batched-tokens 32768 \
  --max-model-len 262144 \
  --reasoning-parser qwen3 \
  --enable-auto-tool-choice \
  --tool-call-parser qwen3_coder

vllm bench serve \
  --backend openai-chat \
  --base-url http://127.0.0.1:8000 \
  --endpoint /v1/chat/completions \
  --dataset-name custom \
  --dataset-path /home/zlab/workspace/jianc/repo/dflash/cache/humaneval.vllm.jsonl \
  --custom-output-len 4096 \
  --num-prompts 32 \
  --max-concurrency 4 \
  --model Qwen/Qwen3.5-122B-A10B \
  --temperature 0.0 \
  --skip-chat-template
```

## Test Result

- Pushed head: `23002d3f368a5a24641301bc71e4ae15dae89a24`.
- `11 passed, 16 warnings` for focused DFlash/KV-sharing tests.
- `pre-commit run --files` passed for the touched DFlash/KV files, including mypy.
- `py_compile` passed for the touched Python files.
- `git diff --check origin/main...HEAD` and `git diff --check` passed.

Qwen3.5-122B-A10B + `z-lab/Qwen3.5-122B-A10B-DFlash`, normal TP4 launch, 15 speculative tokens, HumanEval custom dataset smoke with 4096 output length:

| Metric | Result |
| --- | ---: |
| Available KV cache memory | 94.35 GiB |
| GPU KV cache size | 5,988,876 tokens |
| Max concurrency at 262,144 tokens/request | 22.85x |
| Successful requests | 32 / 32 |
| Output throughput | 521.33 tok/s |
| Median TPOT | 2.45 ms |
| Mean TPOT | 4.50 ms |
| Acceptance rate | 42.23% |
| Acceptance length | 7.33 |
| Draft tokens | 57,045 |
| Accepted tokens | 24,089 |

The KV capacity result confirms this keeps the shared target/draft KV tensor path instead of splitting effective capacity between separate draft and target KV tensors.

### CODE REVIEW

**Comment by DefinitlyEvil** [comment]
Great work, really appreciated! Hopefully this could be implemented and merged soon.

**Comment by Djordje-Stojanovic** [comment]
Cant wait to see this merged hopefully.

**Comment by repne** [comment]
Hi @jianc99, thank you for the great PR. I've tested it extensively in the past few days without issues, but today against latest main it's not working properly. Over 128k context the process hangs (seemingly) indefinitely or the TTFT is in minutes.

**Comment by jianc99** [comment]
@repne Thanks for reporting the issue! I will take a look and fix it.

**Comment by repne** [comment]
Thanks for the fix, I cannot reproduce it anymore with latest main

File: vllm/model_executor/models/qwen3_dflash.py [hunk 1]
```diff
@@ -116,9 +141,13 @@ def __init__(
             num_kv_heads=self.num_kv_heads,
             cache_config=cache_config,
             quant_config=quant_config,
+            per_layer_sliding_window=sliding_window,
             prefix=f"{prefix}.attn",
             attn_type=attn_type,
         )
+        if sliding_window is not None:
+            # DFlash keeps full KV allocation while using SWA only for compute.
```
Comment (benchislett) [inline vllm/model_executor/models/qwen3_dflash.py]: Why? What's the point?

Comment (benchislett) [reply vllm/model_executor/models/qwen3_dflash.py]: What's stopping us from handling this properly?

File: vllm/model_executor/models/qwen3_dflash.py [hunk 2]
```diff
@@ -116,9 +141,13 @@ def __init__(
...
+            self.attn.sliding_window = None
```
[the remark this answers is not in the document]
Comment (benchislett) [reply vllm/model_executor/models/qwen3_dflash.py]: Yeah this seems really hacky

File: vllm/model_executor/models/qwen3_dflash.py [hunk 3]
```diff
@@ -521,7 +563,9 @@ def __init__(self, *, vllm_config: VllmConfig, prefix: str = ""):
             prefix=maybe_prefix(prefix, "lm_head"),
         )
         self.logits_processor = LogitsProcessor(
-            self.config.draft_vocab_size, scale=logit_scale
+            self.config.draft_vocab_size,
+            scale=logit_scale,
+            soft_cap=getattr(self.config, "final_logit_softcapping", None),
```
Comment (benchislett) [inline vllm/model_executor/models/qwen3_dflash.py]: why is this added?

Comment (benchislett) [reply vllm/model_executor/models/qwen3_dflash.py]: Seems like it's not used in the dflash checkpoint...
https://huggingface.co/z-lab/Qwen3.5-122B-A10B-DFlash/blob/main/config.json

File: vllm/v1/worker/gpu_model_runner.py [hunk 4]
```diff
@@ -4996,13 +4996,16 @@ def _get_eagle3_aux_layers_from_config(self) -> tuple[int, ...] | None:
 
         hf_config = self.speculative_config.draft_model_config.hf_config
 
+        is_dflash = self.speculative_config.method == "dflash"
         layer_ids = getattr(hf_config, "eagle_aux_hidden_state_layer_ids", None)
-        if not layer_ids:
+        if is_dflash or not layer_ids:
```
Comment (benchislett) [inline vllm/v1/worker/gpu_model_runner.py]: what's the point of this change?

File: vllm/v1/worker/gpu_model_runner.py [hunk 5]
```diff
@@ -4996,13 +4996,16 @@ def _get_eagle3_aux_layers_from_config(self) -> tuple[int, ...] | None:
...
             dflash_config = getattr(hf_config, "dflash_config", None)
             if dflash_config and isinstance(dflash_config, dict):
                 layer_ids = dflash_config.get("target_layer_ids")
 
         if layer_ids and isinstance(layer_ids, (list, tuple)):
+            if is_dflash:
+                return tuple(layer_id + 1 for layer_id in layer_ids)
```
Comment (benchislett) [inline vllm/v1/worker/gpu_model_runner.py]: This is a bit complicated with speculators. WIP here: https://github.com/vllm-project/vllm/pull/40727

File: vllm/v1/spec_decode/dflash.py [hunk 6]
```diff
@@ -282,7 +282,26 @@ def build_per_group_and_layer_attn_metadata(
...
+                    common_attn_metadata=causal_cad, draft_index=draft_index
+                )
+                for layer_name in causal_layers:
+                    per_layer[layer_name] = attn_metadata
+
         for layer_name, attn_metadata in per_layer.items():
+            if layer_name in sliding_layer_names:
+                assert getattr(attn_metadata, "causal", None) is True, (
+                    f"Attention metadata for sliding layer {layer_name} does not have"
+                    " causal support, which is required for DFlash SWA."
```
Comment (benchislett) [inline vllm/v1/spec_decode/dflash.py:439]: Wait, DFlash SWA is causal? How/why?

File: vllm/model_executor/models/qwen3_dflash.py (hunk 3 above)
Comment (jianc99) [reply vllm/model_executor/models/qwen3_dflash.py]: Oh, this is for the incoming gemma4 draft model. During training we directly borrow the lm_head and embedding from gemma4 target model, which uses embedding scaling and logits soft capping. It's not related to the Qwen3.5-122B model. I will open another PR for this feature.

File: vllm/v1/spec_decode/dflash.py (hunk 6 above)
Comment (jianc99) [reply vllm/v1/spec_decode/dflash.py:439]: Yes, the SWA layers are made causal mainly for compatibility with existing attention backends, since some backends have poor support for non-causal SWA. DFlash is still block diffusion drafting; this is just an implementation choice for the SWA mask. I also verified empirically that, for single-step diffusion drafting, causal SWA performs very similarly to non-causal SWA.

**Comment by jianc99** [comment]
@benchislett Thanks for the review! I cleaned up the PR and force-pushed a smaller stack.

Main changes since the previous version:

- Removed the Gemma4 logits soft-cap change from this PR. I will keep Gemma4 support in a separate PR.
- Removed the `gpu_model_runner.py` layer-id change from this PR. This branch is now stacked on #40727 for DFlash auxiliary layer indexing.
- Removed the `self.attn.sliding_window = None` mutation. DFlash now keeps SWA visible to the compute path and uses a DFlash-specific attention wrapper to return a full-KV cache spec for allocation.
- Reworked DFlash SWA metadata construction so `per_group` and `per_layer` metadata are consistent.
- Updated the unit test to check both group-level and layer-level causal/non-causal metadata.

I also updated the PR description to match the new scope.

File: vllm/model_executor/models/qwen3_dflash.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (jianc99) [reply vllm/model_executor/models/qwen3_dflash.py]: Agreed, this was too hacky. I removed the mutation.

The latest version keeps `sliding_window` on the attention layer so the compute path still sees SWA. To keep full KV allocation for DFlash, I added a small `DFlashAttention` wrapper that converts the returned `SlidingWindowSpec` into a `FullAttentionSpec` while preserving the `sliding_window` value. So SWA remains visible for metadata/backend selection, but the KV allocator does not drop old blocks for the DFlash draft cache.

File: vllm/model_executor/models/qwen3_dflash.py (hunk 3 above)
Comment (jianc99) [reply vllm/model_executor/models/qwen3_dflash.py]: Removed from this PR. That change is for the incoming Gemma4 DFlash checkpoint and is unrelated to Qwen3.5 DFlash SWA, so I split it out into a separate branch/PR.

File: vllm/v1/worker/gpu_model_runner.py (hunk 4 above)
Comment (jianc99) [reply vllm/v1/worker/gpu_model_runner.py]: Removed from this PR. This branch is now stacked on #40727, so the DFlash auxiliary layer-id indexing fix stays there instead of being mixed into the SWA change.

**Comment by repne** [comment]
Hi, thank you for the update. Unfortunately, acceptance rate dropped to 0% for Qwen3.6-27B, and the commit amends + force push + very old main branch make it very hard to bisect my way into debugging the regression so I cannot be of help.

**Comment by jianc99** [comment]
@repne Sorry about that. The zero acceptance issue should be fixed now.

File: vllm/config/speculative.py [hunk 7]
```diff
@@ -1047,6 +1055,10 @@ def use_gemma4_mtp(self) -> bool:
     def use_eagle(self) -> bool:
         return self.method in ("eagle", "eagle3", "mtp", "dflash")
 
+    def requires_eagle_cache_drop(self) -> bool:
```
Comment (benchislett) [inline vllm/config/speculative.py:1058]: This can be taken off the config and inlined directly into the scheduler check. Seems like it's not used anywhere else

File: vllm/model_executor/models/qwen3_dflash.py [hunk 8]
```diff
@@ -47,6 +52,53 @@
 logger = init_logger(__name__)
 
 
+_DFLASH_VALID_LAYER_TYPES = frozenset({"full_attention", "sliding_attention"})
+
+
+def _get_dflash_layer_types(config: Qwen3Config) -> tuple[str, ...]:
```
Comment (benchislett) [inline vllm/model_executor/models/qwen3_dflash.py:58]: Is all this really necessary? Can we lean on existing implementations e.g. Qwen's modelling to implement some of the attention layer boilerplate? Sliding-window is generally well-supported in vLLM

**Comment by dmitriy-sqrt** [comment]
Hey, any chance conflicts could be resolved here, so us - mere mortals can try using it with spark-vllm-docker? :pray:

**Comment by benchislett** [comment]
This feature is being tracked here: https://github.com/vllm-project/vllm/issues/46105

File: vllm/v1/spec_decode/dflash.py [hunk 9]
```diff
@@ -65,9 +71,41 @@ def __init__(
             self.max_positions + 1, device=device, dtype=torch.int32
         )
 
-        # For DFlash we use the input embeddings to embed the mask token
+        # DFlash embeds mask tokens directly.
         self.parallel_drafting_hidden_state_tensor = None
 
+    @override
+    def allow_multiple_draft_kv_cache_groups(self) -> bool:
```
Comment (WindChimeRan) [inline vllm/v1/spec_decode/dflash.py:78]: Since this PR widens unconditionally, do the draft layers ever actually land in **>1 group** here? It looks
like **no** by construction:

`_draft_layer_to_kv_cache_gid` is built by iterating `kv_cache_config.kv_cache_groups` — the **post-widening** groups — so for a uniform DFlash drafter every draft layer maps to the same gid. 

If that's right, then `allow_multiple_draft_kv_cache_groups` (→ could be False), `_draft_layer_to_kv_cache_gid`, and the per-gid block tables / slot mappings are unexercised, and the PR could shrink toward **widening + per-layer-causal + cache-drop + aux-ids**

**Comment by f291400** [comment]
When will this PR be merged? Please merge it ASAP, thanks.

**Comment by jschmied** [comment]
I did a backport to main vllm 0.25.0, see: 

https://github.com/jschmied/vllm/commit/72bfbcf53f6633b012fdec141cb1aae40f0a3fa2

also added tests and guards

dont know if its in shape to commit a PR, comment welcome.

Tested om DGX Spark with qwen36-27b-nvfp4-unsloth and zlab drafter

```
  vllm serve /opt/llm/models/qwen36-27b-nvfp4-unsloth \
    --served-model-name qwen36-27b \
    --kv-cache-dtype fp8 --attention-backend flashinfer \
    --max-model-len 262144 --max-num-seqs 4 --max-num-batched-tokens 8192 \
    --enable-chunked-prefill --enable-prefix-caching \
    --kv-cache-memory-bytes 42949672960 \
    --compilation-config '{"cudagraph_mode":"PIECEWISE","cudagraph_capture_sizes":[1,2,4,8]}' \
    --gpu-memory-utilization 0.9 --load-format fastsafetensors \
    --override-generation-config '{"temperature":0.6,"top_p":0.95,"top_k":20,"min_p":0}' \
    --speculative-config '{"method":"dflash","model":"/opt/llm/models/qwen36-27b-dflash","num_speculative_tokens":8}'
```

**Comment by jschmied** [comment]
Follow-up to my backport comment above — this PR looks superseded, and the difference in approach is worth recording for anyone who lands here.

**Hybrid SWA/full DFlash drafters are already on main**: #47914 ("[Spec Decode] Support hybrid (SWA + full attention) DFlash drafters", 0d12618e9), plus the follow-up #48113. It gates them on the V2 model runner (`VLLM_USE_V2_MODEL_RUNNER=1`) and carries per-layer causality into the attention metadata.

**The two approaches differ in one load-bearing way.** This PR widens a sliding draft layer's KV spec to `FullAttentionSpec` (so DFlash's prewritten context K/V can't be evicted) while the layer keeps its compute window. main doesn't: its sliding draft layers keep `per_layer_sliding_window`, so they get a real `SlidingWindowSpec` and land in a separate KV cache group on their own.

That matters on FlashInfer. Attention groups are keyed on `(backend, kv_cache_spec, num_heads_q)`, so once the KV spec is widened it no longer distinguishes a windowed draft layer from a full one, and they collapse into one group. FlashInfer plans a single prefill wrapper per group and requires a uniform `window_left`, so it refuses ("Window left is not the same for all layers"); and where the keys collide outright, one group's layers are silently dropped, never get a KV cache view, and reach `do_kv_cache_update` as raw 1-D tensors. FlashAttention tolerates per-layer windows, which is presumably why this never showed up here — but FlashInfer is the default attention backend on Blackwell.

My backport (linked above, on v0.25.0 / the V1 runner) fixes that by keying attention groups on the layer's sliding window as well, which is a no-op for any model whose sliding layers already carry a `SlidingWindowSpec` — i.e. exactly the situation main's approach produces. **So this is not a bug in main, and nothing here needs porting forward.** main's design avoids the problem by construction. Worth knowing only if you are on the 0.25 release / V1 runner, where the mixed-attention drafter is otherwise unusable.

Given #47914, I'd suggest this PR can be closed (and #46105 updated) unless there's a reason to keep the V1 path alive.

One thing I would still raise independently of either approach: a mis-wired drafter fails **silently**. My first (wrong) multi-group attempt booted, served, and produced entirely correct output while accepting ~0.3% of drafts — because drafts are verified against the target, a broken drafter never corrupts text, it only stops being accepted, and the sole symptom is that serving is mysteriously slower than no speculation at all. A startup error when a drafter's layers span multiple KV cache groups without the per-group plumbing registered (rather than proceeding) would save the next person a lot of time.
